### PR TITLE
API improvements for discrete

### PR DESCRIPTION
**Pull request by dellaert**
In this PR I made a number of improvements in the API, mainly to make it easy for undergrads in python:

- a new class `DiscretePrior`
- new constructors in `DecisionTreeFactor` and `DiscreteConditional`
- a new `DiscreteConditional::likelihood` method
- `chooseAsFactor` is now just `choose`, ADT version is internal

**update**:
- a new method `DecisionTreeFactor::enumerate`
- a single-value `DiscreteConditional::sample`
- Bumped version to 4.2a0, will release as pre-release after merge

### CODE REVIEW

File: gtsam/discrete/DecisionTreeFactor.cpp [hunk 1]
```diff
@@ -134,17 +134,33 @@ namespace gtsam {
...
   }
 
+  /* ************************************************************************* */
+  std::vector<std::pair<DiscreteValues, double>> DecisionTreeFactor::enumerate() const {
+    // Get all possible assignments
+    std::vector<std::pair<Key, size_t>> pairs;
+    for (auto& key : keys()) {
+      pairs.emplace_back(key, cardinalities_.at(key));
+    }
+    std::vector<std::pair<Key, size_t>> rpairs(pairs.rbegin(), pairs.rend());
```
Comment (ProfFan) [inline gtsam/discrete/DecisionTreeFactor.cpp:144]: Could you explain why it has to be reversed?

File: gtsam/discrete/DiscreteConditional.cpp [hunk 2]
```diff
@@ -97,45 +97,90 @@ bool DiscreteConditional::equals(const DiscreteFactor& other,
...
+                                       const DiscreteValues& parentsValues) {
+  // Get the big decision tree with all the levels, and then go down the
+  // branches based on the value of the parent variables.
+  DiscreteConditional::ADT adt(conditional);
+  size_t value;
+  for (Key j : conditional.parents()) {
+    try {
+      value = parentsValues.at(j);
+      adt = adt.choose(j, value);  // ADT keeps getting smaller.
+    } catch (exception&) {
```
Comment (ProfFan) [inline gtsam/discrete/DiscreteConditional.cpp:110]: use explicit exception type

File: gtsam/discrete/DiscreteConditional.cpp [hunk 3]
```diff
@@ -203,7 +248,7 @@ size_t DiscreteConditional::sample(const DiscreteValues& parentsValues) const {
   static mt19937 rng(2);  // random number generator
 
   // Get the correct conditional density
-  ADT pFS = choose(parentsValues);  // P(F|S=parentsValues)
+  ADT pFS = Choose(*this, parentsValues);  // P(F|S=parentsValues)
 
   // TODO(Duy): only works for one key now, seems horribly slow this way
   assert(nrFrontals() == 1);
```
Comment (ProfFan) [inline gtsam/discrete/DiscreteConditional.cpp:254]: Is this still true?

File: gtsam/discrete/discrete.i [hunk 4]
```diff
@@ -62,20 +77,45 @@ virtual class DiscreteConditional : gtsam::DecisionTreeFactor {
...
-  void add(const gtsam::DiscreteKey& key,
-           const gtsam::DiscreteKeys& parents, string spec);
+  void add(const gtsam::DiscreteConditional& s);
+  void add(const gtsam::DiscreteKey& key, string spec);
+  void add(const gtsam::DiscreteKey& key, string spec,
+           const gtsam::DiscreteKey& parent1);
+  void add(const gtsam::DiscreteKey& key, string spec,
+           const gtsam::DiscreteKey& parent1,
+           const gtsam::DiscreteKey& parent2);
+  void add(const gtsam::DiscreteKey& key, const gtsam::DiscreteKeys& parents,
```
Comment (ProfFan) [inline gtsam/discrete/discrete.i:117]: Why is this different?

File: python/gtsam/tests/test_DiscreteBayesNet.py [hunk 5]
```diff
@@ -53,24 +53,18 @@ def test_Asia(self):
...
 
-        asia.add(Tuberculosis, P([Asia]), "99/1 95/5")
-        asia.add(LungCancer, P([Smoking]), "99/1 90/10")
-        asia.add(Bronchitis, P([Smoking]), "70/30 40/60")
+        asia.add(Tuberculosis, "99/1 95/5", Asia)
+        asia.add(LungCancer, "99/1 90/10", Smoking)
+        asia.add(Bronchitis, "70/30 40/60", Smoking)
 
-        asia.add(Either, P([Tuberculosis, LungCancer]), "F T T T")
+        asia.add(Either, "F T T T", Tuberculosis, LungCancer)
```
Comment (ProfFan) [inline python/gtsam/tests/test_DiscreteBayesNet.py:64]: Test the constructor with array of tuples?

**Review by ProfFan — APPROVED** [review]
Need some clarifications before merge.

File: gtsam/discrete/discrete.i (hunk 4 above)
Comment (ProfFan) [reply gtsam/discrete/discrete.i:117]: * the signature

File: gtsam/discrete/DecisionTreeFactor.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam/discrete/DecisionTreeFactor.cpp:144]: Added comment `// Reverse to make cartesianProduct output a more natural ordering.`

File: gtsam/discrete/DiscreteConditional.cpp (hunk 2 above)
Comment (dellaert) [reply gtsam/discrete/DiscreteConditional.cpp:110]: Done

File: gtsam/discrete/DiscreteConditional.cpp (hunk 3 above)
Comment (dellaert) [reply gtsam/discrete/DiscreteConditional.cpp:254]: This is true. I replaced this with an exception, though.

File: gtsam/discrete/discrete.i (hunk 4 above)
Comment (dellaert) [reply gtsam/discrete/discrete.i:117]: backwards compatible. I am now rethinking this: I'll make the specialization also specify the spec as the last argument.

File: python/gtsam/tests/test_DiscreteBayesNet.py (hunk 5 above)
Comment (dellaert) [reply python/gtsam/tests/test_DiscreteBayesNet.py:64]: Done

**Pull request merged**